### PR TITLE
s3: type corrections and fixes

### PR DESCRIPTION
**Pull request by muvaf**
### Description of your changes

I have tested most of the subresources manually and fixed the errors as I go. I couldn't get to success in some of them like replication configuration due to either me not knowing how to configure them properly or https://github.com/crossplane/provider-aws/issues/383

I believe we should do a patch release after this PR is merged.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Manually.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/s3/v1beta1/notificationConfiguration_types.go
Comment (hasheddan) [inline apis/s3/v1beta1/notificationConfiguration_types.go:115]: Is this field not required? Should it be `*string`?

File: apis/s3/v1beta1/lifecycleConfiguration_types.go
Comment (hasheddan) [inline apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: @muvaf this would be a breaking change and I think would necessitate a minor release, rather than a patch.

**Comment by krishchow** [comment]
Working on some changes, I'll open a PR in your fork @muvaf :)

File: apis/s3/v1beta1/lifecycleConfiguration_types.go
Comment (muvaf) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: @hasheddan Do you think we should bump it to `v1beta2`? Also, I can separate that commit from the others and save jsontag changes for the major release.

Comment (hasheddan) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: Separating commits sounds good :) And I think the correct versioning would be to bump to `v1beta2`. I wonder if this would be a good opportunity to experiment with conversion webhooks...

Comment (muvaf) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: I think we can get away with providing manual instructions for now per [Kubernetes versioning](https://github.com/kubernetes/community/blob/3302b37/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions).

Comment (muvaf) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: So I went ahead and removed the actual breaking changes and the diff is as following:
image

However, I realized that the sub-resources that those fields belong didn't really work in the last release anyway. So, while technically it's more correct to bump to `v1beta2`, in practice I feel like we won't be breaking anyone using `Bucket` resource if we changed those fields and did a patch release. @hasheddan what do you think?

Comment (muvaf) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: @hasheddan pinging

Comment (hasheddan) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: Hmm I am not entirely sure the current behavior of these subresources, my only reservation would be if this could cause a case where a currently valid S3 bucket instance now failed (i.e. if we change field names that are required then a valid instance would become invalid). Since we don't have webhooks yet, maybe we should keep as `v1beta1` but just do a `v0.14` release?

File: apis/s3/v1beta1/notificationConfiguration_types.go
Comment (hasheddan) [reply apis/s3/v1beta1/notificationConfiguration_types.go:115]: following up here

File: apis/s3/v1beta1/lifecycleConfiguration_types.go
Comment (muvaf) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: > if this could cause a case where a currently valid S3 bucket instance now failed

A valid working bucket wouldn't fail because in order to be working, it shouldn't include those fields anyway. If it did, then it's failing because of the errors fixed.

Comment (muvaf) [reply apis/s3/v1beta1/lifecycleConfiguration_types.go:49]: @hasheddan Oh by the way, a full release sounds good.

**Comment by muvaf** [comment]
@krishchow merged your commit into this PR. @hasheddan all your feedback should be addressed.

**Comment by krishchow** [comment]
Did a second pass, LGTM :)

**Comment by muvaf** [comment]
@hasheddan rebased. Can you take a final look? I think we're good to go for a release after this is merged.

**Review by hasheddan — APPROVED** [review]
LGTM, thanks @muvaf! Looks like DCO may be missing on a commit

**Pull request merged**
